### PR TITLE
[BEAM-6627] Added Metrics API processing time reporting to TextIOIT

### PR DESCRIPTION
**Pull request by mwalenia**
* Refactored TimeMonitor to accept any type instead of KV
* Added option to send time metrics to BigQuery

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by mwalenia** [comment]
@lgajowy , could you take a look? Thanks!

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOITHelper.java [hunk 1]
```diff
@@ -65,6 +71,36 @@ public void processElement(ProcessContext c) {
     }
   }
 
+  public static void publishTestMetrics(
```
Comment (lgajowy) [inline sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOITHelper.java]: There's no need to treat time metrics and gcs differently. Both can be read and published using the same set of methods. I think both `publishGcsResults` and `publishTestMetrics` should be refactored to: 

- a method that takes all metrics from the `PipelineResult` and returns `List<NamedTestResult>` 
- a method that logs results to console (this is done in `publishGcsResults` and should be done with other metrics as well
- a method that publishes to BigQuery if bq options are present

So this would give the following method signatures (a general idea): 

1. `Collection<NamedTestResult> readMetrics(PipelineResult result)`
2. `void publishToConsole(Collection<NamedTestResult> results)`
3. `void publishToBigQuery(Collection<NamedTestResult> results)`

For 2 I think you can reuse `ConsoleResultPublisher` from `load-tests` module but please move it to `test-utils) same as you're reusing `BigQueryResultPublisher`.

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOITHelper.java [hunk 2]
```diff
@@ -65,6 +71,36 @@ public void processElement(ProcessContext c) {
...
+      long readEnd = reader.getEndTimeMetric(readEndName);
+      double writeTime = (writeEnd - writeStart) / 1000.0;
+      double readTime = (readEnd - readStart) / 1000.0;
+
+      String uuid = UUID.randomUUID().toString();
+      Timestamp timestamp = Timestamp.now();
+      BigQueryResultsPublisher publisher =
+          BigQueryResultsPublisher.create(bigQueryDataset, NamedTestResult.getSchema());
+      publisher.publish(
+          NamedTestResult.create(uuid, timestamp.toString(), "read_time", readTime), bigQueryTable);
```
Comment (lgajowy) [inline sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOITHelper.java]: Great that you're collecting both read and write time separately. Metrics API FTW! :)

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOITHelper.java [hunk 3]
```diff
@@ -65,6 +71,36 @@ public void processElement(ProcessContext c) {
...
+    if (bigQueryTable != null && bigQueryDataset != null) {
+      MetricsReader reader = new MetricsReader(result, namespace);
+      long writeStart = reader.getStartTimeMetric(writeStartName);
+      long writeEnd = reader.getEndTimeMetric(writeEndName);
+      long readStart = reader.getStartTimeMetric(readStartName);
+      long readEnd = reader.getEndTimeMetric(readEndName);
+      double writeTime = (writeEnd - writeStart) / 1000.0;
+      double readTime = (readEnd - readStart) / 1000.0;
+
+      String uuid = UUID.randomUUID().toString();
```
Comment (lgajowy) [inline sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOITHelper.java]: Please use the same testID for all NamedTestResuts in this test.

*Test id makes selecting all metrics from a particular test run easier in BigQuery. Because of that, the testID should be the same for all metrics in a test invocation.

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java [hunk 4]
```diff
@@ -79,6 +80,7 @@
   private static Integer numShards;
   private static String bigQueryDataset;
   private static String bigQueryTable;
+  private static final String FILEIOIT_NAMESPACE = "fileIOIT";
```
[the remark this answers is not in the document]
Comment (mwalenia) [reply sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java]: Ok, changed it to class name extracted directly from the class

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java [hunk 5]
```diff
@@ -108,13 +110,20 @@ public void writeThenReadAll() {
             .apply(
                 "Produce text lines",
                 ParDo.of(new FileBasedIOITHelper.DeterministicallyConstructTestTextLineFn()))
+            .apply(
+                "Log write start time",
```
[the remark this answers is not in the document]
Comment (mwalenia) [reply sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java]: Sure, changed.

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/metrics/TimeMonitor.java [hunk 6]
```diff
@@ -30,7 +29,7 @@
  * distribution of element processing timestamps, which can be collected and queried using {@link
  * MetricsReader}.
  */
-public class TimeMonitor<K, V> extends DoFn<KV<K, V>, KV<K, V>> {
```
[the remark this answers is not in the document]
Comment (mwalenia) [reply sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/metrics/TimeMonitor.java]: Sure, I extracted it to test-utils.

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOITHelper.java (hunk 3 above)
Comment (mwalenia) [reply sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOITHelper.java]: Ok, done

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOITHelper.java (hunk 1 above)
Comment (mwalenia) [reply sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOITHelper.java]: Refactored, is this what you had in mind?

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOITHelper.java [hunk 7]
```diff
@@ -65,6 +68,13 @@ public void processElement(ProcessContext c) {
     }
   }
 
+  public static void publishToBigQuery(
+      Collection<NamedTestResult> results, String bigQueryDataset, String bigQueryTable) {
+    BigQueryResultsPublisher publisher =
```
Comment (lgajowy) [inline sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOITHelper.java]: There already is a method that suits this case. Could you simplify it even further with: 

`BigQueryResultsPublisher.create().publish() `

using the `BigQueryResutsPublisher.publish(Collection<? extends TestResult> results, String tableName)` method that can accept collection. ([link](https://github.com/apache/beam/blob/67c8e7e6e2ded96b0d854f3b0a43e3ae83c1df6a/sdks/java/testing/test-utils/src/main/java/org/apache/beam/sdk/testutils/publishing/BigQueryResultsPublisher.java#L54))

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java [hunk 8]
```diff
@@ -127,28 +140,49 @@ public void writeThenReadAll() {
...
-    publishGcsResults(result);
+    gatherAndPublishMetrics(result);
   }
 
-  private void publishGcsResults(PipelineResult result) {
+  private void gatherAndPublishMetrics(PipelineResult result) {
+    String uuid = UUID.randomUUID().toString();
+    Timestamp timestamp = Timestamp.now();
+    List<NamedTestResult> namedTestResults = readMetrics(result, uuid, timestamp);
+    publishToBigQuery(namedTestResults, bigQueryDataset, bigQueryTable);
```
Comment (lgajowy) [inline sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java]: We're missing a check here if the `bigQueryDataset` and `bigQueryTable` are present. Could you add one? 
We should not publish to bigQuery if user doesn't explicitly say so (by passing the options).

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java [hunk 9]
```diff
@@ -127,28 +140,49 @@ public void writeThenReadAll() {
...
+          NamedTestResult.create(uuid, timestamp.toString(), "copies_per_sec", copiesPerSec));
+    }
+
+    results.add(NamedTestResult.create(uuid, timestamp.toString(), "read_time", readTime));
+    results.add(NamedTestResult.create(uuid, timestamp.toString(), "write_time", writeTime));
+
+    return results;
+  }
+
+  private double calculateGcsMetric(PipelineResult result) {
```
Comment (lgajowy) [inline sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java]: Could we call this `getGcsCopies` or `calculateGcsCopies` (I'm fine with both)?

Comment (lgajowy) [reply sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java]: More importantly, this should be collected only when `options.getGcsPerformanceMetrics()` is `true`. I think this was missing before. Could you add that condition?

@udim am I right here or am I missing something?

Tl;dr version for Udi: we want to reuse the mechanism for publishing the bigQuery metrics for collecting time in IOITs too. Previously we checked if BigQuery related options are present to publish the Gcs metric. Now that there are more metrics collected this way, such "if" condition would be too broad.

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOITHelper.java (hunk 7 above)
Comment (lgajowy) [reply sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOITHelper.java]: For the future record: we discussed a bit about extracting a base class that would encapsulate collecting + publishing metrics. Such class could be then extended by particular test cases so that only relevant code is there. This would drastically reduce duplication. 

This is way out of scope of this PR. I created an issue for this: https://issues.apache.org/jira/browse/BEAM-6662

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java (hunk 9 above)
Comment (mwalenia) [reply sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java]: Added an option switching GCS metric reporting to file based tests

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java (hunk 8 above)
Comment (mwalenia) [reply sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java]: Right, I missed it. Corrected

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOITHelper.java (hunk 7 above)
Comment (mwalenia) [reply sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOITHelper.java]: Refactored reporting to use`BigQueryResutsPublisher.publish(Collection<? extends TestResult> results, String tableName)` method

**Comment by mwalenia** [comment]
@lgajowy I think it's ready for another review. Thanks for the tips!

**Comment by mwalenia** [comment]
Run Java ManyFilesTextIO Performance Test

**Comment by mwalenia** [comment]
Run Seed Job

**Comment by mwalenia** [comment]
Run Java ManyFilesTextIO Performance Test

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java (hunk 9 above)
Comment (udim) [reply sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java]: I guess the way it was working before was by checking `if (numCopies < 0 || copyTimeMsec < 0)`, which is false if `options.getGcsPerformanceMetrics()` is false.
So I think we can merge --reportGcsPerformanceMetrics and --gcsPerformanceMetrics into the latter. No need to add a separate flag, and only report copiesPerSec if --gcsPerformanceMetrics is set.

Comment (mwalenia) [reply sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java]: Thanks for the input! You're right, that would be optimal. However, I don't see a way to access this switch from inside the test - it's a part of `GcsOptions` class. Trying to use an option with the same name in two places leads to compilation problems (two clashing method definitions). Is there another, better option that you know of?

Comment (udim) [reply sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java]: Have you tried: `options.as(GcsOptions.class).getGcsPerformanceMetrics()`?

Comment (mwalenia) [reply sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java]: I didn't try this, thanks for the suggestion! I changed the test so that the additional flag is not necessary.

**Comment by mwalenia** [comment]
Run Seed Job

**Comment by mwalenia** [comment]
Run Java ManyFilesTextIO Performance Test

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOTestPipelineOptions.java [hunk 10]
```diff
@@ -54,4 +54,10 @@
   String getBigQueryTable();
 
   void setBigQueryTable(@Nullable String tableName);
+
+  @Description("Option to gather GCS performance metrics")
+  @Default.Boolean(false)
+  boolean getReportGcsPerformanceMetrics();
```
Comment (udim) [inline sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOTestPipelineOptions.java]: Don't forget to remove this flag if you don't need it.

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java (hunk 9 above)
Comment (lgajowy) [reply sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java]: Interesting: if I read this correctly, we didn’t have to import the `sdks-java-extensions-google-cloud-platform` dependency because it is a transitive dependency from test-utils package. What if we refactor test utils and get rid of the extensions there? If we stick with one flag, we should import the `sdks-java-extensions-google-cloud-platform` dependency in `file-based-io-tests` module.

Looking at this in a broader, more future-proof way, I’ve begun to think it's better to stick with 2 separate flags: first flag in the production code that makes the IO/pipeline/whatever else actually produce the metric and the other one in the test code that explicitly tells that the metric should be harvested. IMO it has the following benefits: 

 - we separate the concerns of producing and collecting the metrics - people reading test code/implementing new tests can find/define the latter option in TestPipelineOptions interface, read it’s the description and decide if it should be used without digging into beam implementation specifics and searching for the option in the whole codebase 
 - we don't have to import runner/IO specific dependencies every time similar case occurs (like: ‘sdks-java-extensions-google-cloud-platform`) - this is especially important in case we’re running on other runners than Dataflow. I can imagine that tests run on Flink may not require dataflow/gcs related dependencies

The disadvantage is that we have 2 flags for each case like this but I think such solution will “scale” better in case we require more flags like this from other filesystems/runners (eg s3 copies/hdfs copies).

wdyt?

Comment (udim) [reply sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java]: @lgajowy I think you're right that separating the setting of collection and reporting makes more sense, especially considering build dependencies.
For metric reporting, a single flag in the style of --experiments might simplify things (no need to add a new flag every time a metric is added).

Comment (mwalenia) [reply sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java]: @udim What about cases in which we need some metrics, but not others? Take for example the hdfs copies, when we want this metric, other may be irrelevant, like gs copies. In that case we'd need to check if each metric has valid values and then decide whether to report it (like it's done now with gcs copies - number of operations greater than 0 and time not below 0? Calculate and report). I believe setting explicit flags for every metric wouldn't be simpler, but it would be easier to precisely control what is gathered and reported.

**Comment by mwalenia** [comment]
Run Seed Job

**Comment by mwalenia** [comment]
Run Java ManyFilesTextIO Performance Test

**Comment by mwalenia** [comment]
@lgajowy I rolled back the changes to test options, I think it's ready for review now.

File: sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java (hunk 9 above)
Comment (udim) [reply sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java]: @mwalenia Explicit flags for each metric isn't simpler, but perhaps using just 2 flags:
```
--metricsGather=hdfsCopies,gcsCopies,...
--metricsReportToBigQuery=hdfsCopies,gcsCopies,...
```

Comment (mwalenia) [reply sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java]: @udim A big advantage of using my approach is built-in error reporting and checking. After passing an argument with a typo we get for example:

`java.lang.IllegalArgumentException: Class interface org.apache.beam.sdk.testing.TestPipelineOptions missing a property named 'compresionType'. Did you mean 'compressionType'?`

whereas there's no such mechanism when passing a list of strings as parameter value. In case of a typo we'd have a silent failure, since in order to switch reporting of different metrics on and off, we'd need to search for strings in an array.

I think we should stick with explicitly set flags for subsequent metrics. WDYT?

Comment (udim) [reply sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/text/TextIOIT.java]: Sounds good!

**Comment by mwalenia** [comment]
@lgajowy We resolved the discussion in comments, it's ready for review

**Comment by lgajowy** [comment]
Run Java Load Tests Smoke

**Comment by lgajowy** [comment]
Run Java Load Tests Smoke

**Comment by lgajowy** [comment]
Run seed job

**Comment by lgajowy** [comment]
Run Java TextIO Performance Test HDFS

**Comment by lgajowy** [comment]
Run Java CompressedTextIO Performance Test HDFS

**Comment by lgajowy** [comment]
Run Java ManyFilesTextIO Performance Test HDFS

**Comment by lgajowy** [comment]
Run Java TextIO Performance Test

**Comment by lgajowy** [comment]
Run Java CompressedTextIO Performance Test

**Comment by lgajowy** [comment]
Run Java ManyFilesTextIO Performance Test

**Comment by lgajowy** [comment]
Run seed job

**Comment by lgajowy** [comment]
Run Java TextIO Performance Test

**Comment by lgajowy** [comment]
Run Java CompressedTextIO Performance Test

**Comment by lgajowy** [comment]
Run Java ManyFilesTextIO Performance Test

**Comment by lgajowy** [comment]
Great! LGTM. 

merging despite failing Java PreCommit. Failure is unrelated to this change and is seen in all Prs now. More details here: [BEAM-6779](https://issues.apache.org/jira/browse/BEAM-6779)
